File: python/3429. Paint House IV.py
class Solution:
    def minCost(self, n: int, cost: List[List[int]]) -> int:
        @cache
        def find_cost(idx, left, right):
            if idx >= n/2:
                return 0
            answer = float('inf')
            for i in range(3):
                for j in range(3):
                    if i == left or j == right or i == j:
                        continue
                    answer = min(answer, cost[idx][i]+cost[-idx-1][j]+find_cost(idx+1, i, j))
            return answer
        return find_cost(0, -1, -1)

File: java/3429. Paint House IV.java
class Solution {
    public long minCost(int n, int[][] cost) {
        Long[][][] cache = new Long[n][4][4];
        return findCost(n, cost, 0, 0, 0, cache);
    }
    private long findCost(
        int n,
        int[][] cost,
        int idx,
        int left,
        int right,
        Long[][][] cache
    )
    {
        if (idx >= n/2) return 0;
        long answer = Long.MAX_VALUE;
        if (cache[idx][left][right] != null) return cache[idx][left][right];
        for (int i = 1; i <= 3; i++) {
            for (int j = 1; j <= 3; j++) {
                if (i == left || j == right || i == j) continue;
                long val = findCost(idx+1, i, j)
                if (val != Long.MAX_VALUE) {
                    answer = Math.min(answer, cost[idx][i-1] + cost[n-1-idx][j-1] + val);
                }
            }
        }
        return cache[idx][left][right] = answer;
    }
}